hermes/crates/cardano-chain-follower/src/lib.rs

```diff
@@ -1,6 +1,213 @@
+//!
+//!     let _update = follower.next().await.unwrap();
+//! }
+//! ```
 
-/// Greets a name.
-pub fn greet(name: &str) {
-    println!("Hello, {name}!");
+// TODO: remove this once we implement the API.
+#![allow(dead_code, clippy::unused_async)]
+// (fsgr): This should be removed. I only added it because, for some reason,
+//         the tower crate is failing to compile in my machine (didn't test anywhere else)
+//         if it's compiled with this flag.
+#![deny(missing_docs)]
+
+pub use pallas::network::miniprotocols::Point;
+use pallas::{
+    ledger::traverse::MultiEraBlock,
+    network::{
+        facades::{NodeClient, PeerClient},
+        miniprotocols::{MAINNET_MAGIC, PREVIEW_MAGIC, PRE_PRODUCTION_MAGIC, TESTNET_MAGIC},
+    },
+};
+
+/// Crate error type.
+///
+/// We are using a boxed error here until we have some implementation of the
+/// the crate's API.
+///
+/// In the future this will probably be something as:
+///
+/// ```ignore
+/// use thiserror::Error;
+///
+/// #[derive(Debug, Error)]
+/// pub enum Error {
+/// ...
+/// }
+/// ```
+pub type Error = Box<dyn std::error::Error>;
+
+/// Crate result type.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// A point in the chain or the tip.
+pub enum PointOrTip {
+    /// Represents a specific point of the chain.
+    Point(Point),
+    /// Represents the tip of the chain.
+    Tip,
+}
+
+impl From<Point> for PointOrTip {
+    fn from(point: Point) -> Self {
+        Self::Point(point)
+    }
+}
+
+/// CBOR encoded data of a multi-era block.
+pub struct MultiEraBlockData(Vec<u8>);
+
+impl MultiEraBlockData {
+    /// Decodes the data into a multi-era block.
+    ///
+    /// # Errors
+    ///
+    /// Returns Err if the block's era couldn't be decided or if the encoded data is
+    /// invalid.
+    pub fn decode(&self) -> Result<MultiEraBlock> {
+        let block = MultiEraBlock::decode(&self.0).map_err(Box::new)?;
+
+        Ok(block)
+    }
+}
+
+/// Enum of possible Cardano networks.
+pub enum Network {
+    /// Cardano mainnet network.
+    Mainnet,
+    /// Cardano pre-production network.
+    Preprod,
+    /// Cardano preview network.
+    Preview,
+    /// Cardano testnet network.
+    Testnet,
+}
+
+impl From<Network> for u64 {
+    fn from(network: Network) -> Self {
+        match network {
+            Network::Mainnet => MAINNET_MAGIC,
+            Network::Preprod => PRE_PRODUCTION_MAGIC,
+            Network::Preview => PREVIEW_MAGIC,
+            Network::Testnet => TESTNET_MAGIC,
+        }
+    }
+}
+
+/// Enum of possible types of clients used by the follower.
+enum Client {
+    /// Pallas node-to-node client.
+    N2n(PeerClient),
+    /// Pallas node-to-client client.
+    N2c(NodeClient),
+}
+
+/// Enum of chain updates received by the follower.
+pub enum ChainUpdate {
+    /// New block inserted on chain.
+    Block(MultiEraBlockData),
+    /// Chain rollback to the given point.
+    Rollback(Point),
```

Comment:
Not sure about `Rollback(Point)` it is possible to change it to `Rollback(MultiEraBlockData)` ?

Reply:
Yeah that would be possible but we need to figure out fetching blocks for both node-to-node and node-to-client connections (because the node just returns the point in case of a rollback).